src/anml/data/data.py

```diff
@@ -144,3 +148,31 @@ def process_data(self, df: pd.DataFrame):
                 )
             if not self.multi_spec:
                 self.data[name] = self.data[name][0]
+
+    def process_params(self, df: pd.DataFrame):
+        """Process a data frame's covariates and attach to this instance with existing parameter sets.
+
+        Parameters
+        ----------
+        df
+            A pandas.DataFrame with all of the information that the existing parameter sets encode
+            for the covariates.
+
+        """
+        if len(self._param_set) == 0:
+            raise EmptySpecsError("Need to attach parameter sets before processing data.")
+
+        for param_set in self._param_set:
+            param_set._validate_df(df=df)
+
+        for param_set in self._param_set:
+            for parameter in param_set.parameters:
+                covariates = dict()
+                covariates[parameter.param_name] = list()
+                for variable in parameter.variables:
+                    covariates[parameter.param_name].append(
+                        variable.design_mat(df=df)
+                    )
```

Comment:
is the `covariates` has only one key (i.e. `parameter.param_name`)? (just checking my understanding :)

Reply:
Yes!